cmd/checkpoint/main.go

```diff
@@ -19,234 +21,499 @@ import (
+			//       However, we don't know this, and as far as the checkpointer is concerned - that pod is no longer scheduled to this node.
+			if _, ok := apiParentPods[id]; !ok {
+				glog.V(4).Infof("API GC: should remove inactive checkpoint %s", id)
+				removeMap[id] = struct{}{}
+				delete(inactiveCheckpoints, id)
+			}
+		}
+
+		// Scan active checkpoints we should GC
+		for id := range activeCheckpoints {
+			// If the active checkpoint does not have a parent in the api-server, we must assume it should no longer be running on this node.
+			if _, ok := apiParentPods[id]; !ok {
+				glog.V(4).Infof("API GC: should remove active checkpoint %s", id)
+				removeMap[id] = struct{}{}
+				delete(activeCheckpoints, id)
 			}
 		}
-		time.Sleep(1 * time.Second)
 	}
+
+	// Can make decisions about starting/stopping checkpoints just with local state.
+
+	// If there is an inactive checkpoint, and no parent is running, start the checkpoint
+	for id := range inactiveCheckpoints {
+		if _, ok := localParentPods[id]; !ok {
+			glog.V(4).Infof("Should start checkpoint %s", id)
+			start = append(start, id)
+		}
+	}
+
+	// If there is an active checkpoint and a parent pod, stop the active checkpoint
+	// The parent may not be in a running state, but the kubelet is trying to start it so we should get out of the way.
+	for id := range activeCheckpoints {
+		if _, ok := localParentPods[id]; ok {
+			glog.V(4).Infof("Should stop checkpoint %s", id)
+			stop = append(stop, id)
+		}
+	}
+
+	// De-duped checkpoints to remove. If we decide to GC a checkpoint, we will clean up both inactive/active.
+	for k := range removeMap {
+		remove = append(remove, k)
+	}
+
+	return start, stop, remove
 }
 
-func stripNonessentialInfo(p *v1.Pod) {
-	p.Spec.ServiceAccountName = ""
-	p.Spec.DeprecatedServiceAccount = ""
-	p.Status.Reset()
+// Create checkpoint will iterate through pods which are candidates for checkpoing, and if they should be:
+// - checkpoint any remote assets they need (e.g. secrets)
+// - sanitize their podSpec, removing unnecessary information
+// - storing the manifest on disk in an "inactive" checkpoint location
+//TODO(aaron): Add support for checkpointing configMaps
+func createCheckpoints(client clientset.Interface, pods map[string]*v1.Pod) {
+	for _, pod := range pods {
+		if !isRunning(pod) {
+			continue
+		}
+		id := podID(pod)
+
+		cp, err := copyPod(pod)
+		if err != nil {
+			glog.Errorf("Failed to create checkpoint pod copy for %s: %v", id, err)
+			continue
+		}
+
+		cp, err = checkpointSecretVolumes(client, cp)
+		if err != nil {
+			glog.Errorf("Failed to checkpoint secrets for pod %s: %v", id, err)
+			continue
+		}
+		cp, err = sanitizeCheckpointPod(cp)
+		if err != nil {
+			glog.Errorf("Failed to sanitize manifest for %s: %v", id, err)
+			continue
+		}
+		if err := ensureCheckpointManifest(cp); err != nil {
```

Comment:
I used ensure because it may not write the manifest if it already exists and is the same. No strong feelings here, could use `writeCheckpointManifest` or maybe `writeIfNotExistsCheckpointManifest`